From d99a60bf63f6a19f2a102b96c259b378ec8c736f Mon Sep 17 00:00:00 2001
From: Jani Hautakangas <jani@kodegood.com>
Date: Mon, 22 Dec 2025 10:43:02 +0200
Subject: [PATCH 5/7] Meta/CMake: Do not use -march=native when cross-compiling

The introduction of ENABLE_CI_BASELINE_CPU made -march=native the default
for non-CI builds. This breaks cross-compilation with Clang, where native
CPU detection is explicitly unsupported when a target triple is set.

Guard the use of -march=native behind NOT CMAKE_CROSSCOMPILING so native
tuning is only applied for true host builds, while cross builds continue
to use the toolchain defaults.
---
 Meta/CMake/common_compile_options.cmake | 24 ++++++++----------------
 1 file changed, 8 insertions(+), 16 deletions(-)

diff --git a/Meta/CMake/common_compile_options.cmake b/Meta/CMake/common_compile_options.cmake
index db0ca07c47..d89251f97d 100644
--- a/Meta/CMake/common_compile_options.cmake
+++ b/Meta/CMake/common_compile_options.cmake
@@ -49,7 +49,10 @@ macro(add_swift_link_options)
     add_link_options($<$<LINK_LANGUAGE:Swift>:${args}>)
 endmacro()
 
-if (ENABLE_CI_BASELINE_CPU)
+if (MSVC)
+    # On Windows, we always target the same architecture since there is no -march=native equivalent.
+    add_cxx_compile_options(/arch:AVX2)
+elseif (ENABLE_CI_BASELINE_CPU)
     # In CI, we want to target a common architecture so different runners can share ccache caches effectively.
     if (APPLE AND CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
         add_cxx_compile_options(-mcpu=apple-m1)
@@ -58,14 +61,12 @@ if (ENABLE_CI_BASELINE_CPU)
     elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
         add_cxx_compile_options(-march=x86-64-v3)
     endif()
-elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "riscv64")
-    # On RISC-V the generic -march=native is not yet supported and both gcc and clang require an explicit
-    # ISA or target string. Unfortunately hardware probing is also neither easy nor reliable at the moment.
-    # For the time being use the defaults for the best compatibility with existing hardware and toolchains.
-    # FIXME: Remove this branch once -march=native is supported.
 else()
     # In all other cases, compile for the native architecture of the host system.
-    add_cxx_compile_options(-march=native)
+    # This is only valid when not cross-compiling.
+    if (NOT CMAKE_CROSSCOMPILING)
+        add_cxx_compile_options(-march=native)
+    endif()
 endif()
 
 # FIXME: Rework these flags to remove the suspicious ones.
@@ -135,10 +136,6 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND NOT CMAKE_CXX_SIMULATE_ID  MATCHES
     add_cxx_compile_options(-Wno-implicit-const-int-float-conversion)
     add_cxx_compile_options(-Wno-user-defined-literals)
     add_cxx_compile_options(-Wno-unqualified-std-cast-call)
-
-    # Used for the #embed directive.
-    # FIXME: Remove this once #embed is no longer an extension.
-    add_cxx_compile_options(-Wno-c23-extensions)
 elseif (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     # Only ignore expansion-to-defined for g++, clang's implementation doesn't complain about function-like macros
     add_cxx_compile_options(-Wno-expansion-to-defined)
@@ -158,11 +155,6 @@ elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang$" AND CMAKE_CXX_SIMULATE_ID MATCHES
     add_cxx_compile_options(-Wno-reserved-identifier)
     add_cxx_compile_options(-Wno-user-defined-literals)
     add_cxx_compile_options(-Wno-unqualified-std-cast-call)
-
-    # Used for the #embed directive.
-    # FIXME: Remove this once #embed is no longer an extension.
-    add_cxx_compile_options(-Wno-c23-extensions)
-
     add_definitions(-D_CRT_SECURE_NO_WARNINGS)
 
     # TODO: this seems wrong, but we use this kind of code too much
-- 
2.48.1

